pkg/ingester/ingester.go

```diff
@@ -283,6 +370,99 @@ func (i *Ingester) Push(ctx old_ctx.Context, req *client.WriteRequest) (*client.
 	return &client.WriteResponse{}, lastPartialErr
 }
 
+// v2Push adds metrics to a block
+func (i *Ingester) v2Push(ctx old_ctx.Context, req *client.WriteRequest) (*client.WriteResponse, error) {
+	userID, err := user.ExtractOrgID(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("no user id")
+	}
+
+	db, err := i.getOrCreateTSDB(userID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Walk the samples, appending them to the users database
+	app := db.Appender()
+	for _, ts := range req.Timeseries {
+		for _, s := range ts.Samples {
+			if i.stopped {
+				return nil, fmt.Errorf("ingester stopping")
+			}
+			lset := make(lbls.Labels, len(ts.Labels))
+			for i := range ts.Labels {
+				lset[i] = lbls.Label{
+					Name:  ts.Labels[i].Name,
+					Value: ts.Labels[i].Value,
+				}
+			}
+			if _, err := app.Add(lset, s.TimestampMs, s.Value); err != nil {
+				if err := app.Rollback(); err != nil {
+					level.Warn(util.Logger).Log("failed to rollback on error", "userID", userID)
+				}
+				return nil, err
+			}
+		}
+	}
+	if err := app.Commit(); err != nil {
+		return nil, err
+	}
+
+	client.ReuseSlice(req.Timeseries)
+
+	return &client.WriteResponse{}, nil
+}
+
+func (i *Ingester) getTSDB(userID string) *tsdb.DB {
+	i.userStatesMtx.RLock()
+	defer i.userStatesMtx.RUnlock()
+	db, _ := i.dbs[userID]
+	return db
+}
+
+func (i *Ingester) getOrCreateTSDB(userID string) (*tsdb.DB, error) {
+	db := i.getTSDB(userID)
+	if db == nil {
+		i.userStatesMtx.Lock()
+		defer i.userStatesMtx.Unlock()
+		udir := i.userDir(userID)
+
+		// Create a new user database
+		var err error
+		db, err = tsdb.Open(udir, util.Logger, nil, &tsdb.Options{
+			RetentionDuration: uint64(i.retention / time.Millisecond),
+			BlockRanges:       []int64{int64(i.blockRanges / time.Millisecond)},
+			NoLockfile:        true,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		// Create a new shipper for this database
+		l := lbls.Labels{
+			{
+				Name:  "user",
+				Value: userID,
+			},
+		}
+		s := shipper.New(util.Logger, nil, udir, &Bucket{userID, i.bucket}, func() lbls.Labels { return l }, metadata.ReceiveSource)
+		i.done.Add(1)
+		go func() {
+			defer i.done.Done()
+			runutil.Repeat(30*time.Second, i.quit, func() error {
+				if uploaded, err := s.Sync(context.Background()); err != nil {
```

Comment:
I think it should probably be a context with cancel and not a timeout to shutdown more gracefully. Since it's scanning blocks it can take an arbitrary amount of time, and if it fails to get through it all once, it likely won't be able to do it again faster (since more work would be accumulating), leaving it in a state where it can never complete a sync.

What do you think?

Reply:
I do like contexts... they can have both a cancel and a timeout. However, if this operation is the whole scan of data, then yes I see how a flat timeout would not work here.

Reply:
I think I'm going to leave as is for now, since using background lets a sync complete during shutdown. We can add a cancel context in the future if shutdowns are taking too long waiting for a sync to complete.